Source/WebCore/rendering/updating/RenderTreeBuilder.cpp

```diff
@@ -904,6 +904,14 @@ void RenderTreeBuilder::updateAfterDescendants(RenderElement& renderer)
         listBuilder().updateItemMarker(downcast<RenderListItem>(renderer));
     if (is<RenderBlockFlow>(renderer))
         multiColumnBuilder().updateAfterDescendants(downcast<RenderBlockFlow>(renderer));
+
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    // If the <svg> has no children, nobody created the anonymous RenderSVGViewportContainer
+    // (usually happens during insertion of first <svg> child into the render tree) - change
+    // that here and create it.
+    if (auto* svgRoot = dynamicDowncast<RenderSVGRoot>(renderer))
+        svgBuilder().createViewportContainerIfNeeded(*svgRoot);
```

Comment:
Hmm, not sure I got your question right. Are you questioning the code itself or where it is located?

Usually the RenderSVGViewportContainer wrapper is created lazily in findOrCreateParentForChild() in the SVG builder. The code here in updateAfterDescendants() only has an effect if the had no children -- otherwise the anonymous wrapper is already created.

Reply:
This is about the location, I think it is better to keep svg stuff in svg builders. Anyway here it looks like we should at least adapt the updateAfterDescendants naming like the other builder calls just above, and do our SVG things in there.

Reply:
Ah I see your point. The comment and the createViewportcontainerIfNeeded() call should be moved into RenderTreeBuilder::SVG::updateAfterDescendants(RenderSVGRoot&) -- this way we don't pollute RenderTreeBuilder with details what SVG/LBSE is doing when updateAfterDescendants() is invoked.

I've also reordered the function, a RenderSVGRoot cannot be a RenderBlock, RenderBlockFlow or RenderListItem, thus we can early return for RenderSVGRoot. Besides that all if() statements should use dynamicDowncast. This is the most concise variant I came up with:

```
void RenderTreeBuilder::updateAfterDescendants(RenderElement& renderer)
{
#if ENABLE(LAYER_BASED_SVG_ENGINE)
    if (auto* svgRoot = dynamicDowncast<RenderSVGRoot>(renderer)) {
        svgBuilder().updateAfterDescendants(*svgRoot);
        return; // A RenderSVGRoot cannot be a RenderBlock, RenderListItem or RenderBlockFlow: early return.
    }
#endif

    // Do not early return here in any case. For example, RenderListItem derives
    // from RenderBlockFlow and indirectly from RenderBlock thus fulfilling all
    // update conditions below.
    if (auto* block = dynamicDowncast<RenderBlock>(renderer))
        firstLetterBuilder().updateAfterDescendants(*block);
    if (auto* listItem = dynamicDowncast<RenderListItem>(renderer))
        listBuilder().updateItemMarker(*listItem);
    if (auto* blockFlow = dynamicDowncast<RenderBlockFlow>(renderer))
        multiColumnBuilder().updateAfterDescendants(*blockFlow);
}
```

Please check the new iteration, once I pushed it :-)